File: Airwallex/Card/Internal/AWXCardViewModel.m
//
//  AWXCardViewModel.m
//  Card
//
//  Created by Hector.Huang on 2022/9/14.
//  Copyright © 2022 Airwallex. All rights reserved.
//

#import "AWXCardViewModel.h"
#import "AWXCard.h"
#import "AWXCardProvider.h"
#import "AWXCardValidator.h"
#import "AWXCountry.h"
#import "AWXSession.h"

@interface AWXCardViewModel ()

@property (nonatomic, readwrite) BOOL isReusingShippingAsBillingInformation;
@property (nonatomic, strong, nonnull) AWXSession *session;

@end

@implementation AWXCardViewModel

- (instancetype)initWithSession:(AWXSession *)session supportedCardSchemes:(NSArray<AWXCardScheme *> *)cardSchemes {
    self = [super init];
    if (self) {
        _session = session;
        _isReusingShippingAsBillingInformation = session.billing != nil && session.isBillingInformationRequired;
        _selectedCountry = [AWXCountry countryWithCode:session.billing.address.countryCode];
        _supportedCardSchemes = cardSchemes;
    }
    return self;
}

- (BOOL)isBillingInformationRequired {
    return self.session.isBillingInformationRequired;
}

- (BOOL)isCardSavingEnabled {
    return [self.session isKindOfClass:[AWXOneOffSession class]] && self.session.customerId;
}

- (AWXPlaceDetails *)initialBilling {
    return self.session.billing;
}

- (BOOL)setReusesShippingAsBillingInformation:(BOOL)reusesShippingAsBillingInformation error:(NSString **)error {
    if (reusesShippingAsBillingInformation && self.session.billing == nil) {
        if (error != NULL) {
            *error = NSLocalizedString(@"No shipping address configured.", nil);
        }

        return NO;
    } else {
        self.isReusingShippingAsBillingInformation = reusesShippingAsBillingInformation;

        return YES;
    }
}

#pragma mark Data creation

- (AWXPlaceDetails *)makeBillingWithFirstName:(NSString *)firstName
                                     lastName:(NSString *)lastName
                                        email:(NSString *)email
                                  phoneNumber:(NSString *)phoneNumber
                                        state:(NSString *)state
                                         city:(NSString *)city
                                       street:(NSString *)street
                                     postcode:(NSString *)postcode {
    if (self.isReusingShippingAsBillingInformation) {
        return self.session.billing.copy;
    }

    AWXPlaceDetails *place = [AWXPlaceDetails new];
    place.firstName = firstName;
    place.lastName = lastName;
    place.email = email;
    place.phoneNumber = phoneNumber;

    AWXAddress *address = [AWXAddress new];
    address.countryCode = self.selectedCountry.countryCode;
    address.state = state;
    address.city = city;
    address.street = street;
    address.postcode = postcode;

    place.address = address;
    return place;
}

- (AWXCard *)makeCardWithName:(NSString *)name
                       number:(NSString *)number
                       expiry:(NSString *)expiry
                          cvc:(NSString *)cvc {
    NSArray *dates = [expiry componentsSeparatedByString:@"/"];

    AWXCard *card = [AWXCard new];
    card.name = name;
    card.number = [number stringByReplacingOccurrencesOfString:@" " withString:@""];
    card.expiryYear = [NSString stringWithFormat:@"20%@", dates.lastObject];
    card.expiryMonth = dates.firstObject;
    card.cvc = cvc;

    return card;
}

- (NSArray *)makeDisplayedCardBrands {
    NSMutableArray *cardBrands = [NSMutableArray new];
    for (id brand in AWXCardSupportedBrands()) {
        for (AWXCardScheme *cardScheme in _supportedCardSchemes) {
            if ([self cardBrandFromCardScheme:cardScheme] == [brand intValue]) {
                [cardBrands addObject:brand];
            }
        }
    }
    return cardBrands;
}

#pragma mark Data validation

- (AWXPlaceDetails *)validatedBillingDetails:(AWXPlaceDetails *)billing error:(NSString **)error {
    AWXPlaceDetails *validated = billing.copy;
    NSString *validationError = [validated validate];
    if (validationError != nil) {
        if (error != NULL) {
            *error = validationError;
        }

        return nil;
    } else {
        return validated;
    }
}

- (AWXCard *)validatedCardDetails:(AWXCard *)card error:(NSString **)error {
    NSString *validationError = [card validate];
    if (validationError != nil) {
        if (error != NULL) {
            *error = validationError;
        }

        return nil;
    } else {
        return card;
    }
}

- (NSString *)validationMessageFromCardNumber:(NSString *)cardNumber {
    if (cardNumber.length > 0) {
        if ([AWXCardValidator.sharedCardValidator isValidCardLength:cardNumber]) {
            NSString *cardName = [AWXCardValidator.sharedCardValidator brandForCardNumber:cardNumber].name;
            for (AWXCardScheme *cardScheme in _supportedCardSchemes) {
                if ([cardScheme.name isEqualToString:cardName.lowercaseString]) {
                    return NULL;
                }
            }
            return NSLocalizedString(@"Card not supported for payment", nil);
        }
        return NSLocalizedString(@"Card number is invalid", nil);
    }
    return NSLocalizedString(@"Card number is required", nil);
}

#pragma mark Payment

- (AWXCardProvider *)preparedProviderWithDelegate:(id<AWXProviderDelegate>)delegate {
    return [[AWXCardProvider alloc] initWithDelegate:delegate session:self.session];
}

- (AWXDefaultActionProvider *)actionProviderForNextAction:(AWXConfirmPaymentNextAction *)nextAction
                                             withDelegate:(id<AWXProviderDelegate> _Nullable)delegate {
    Class class = ClassToHandleNextActionForType(nextAction);
    if (class == nil) {
        return nil;
    }

    AWXDefaultActionProvider *actionProvider = [[class alloc] initWithDelegate:delegate session:self.session];
    return actionProvider;
}

- (BOOL)confirmPaymentWithProvider:(AWXCardProvider *_Nonnull)provider
                           billing:(AWXPlaceDetails *)placeDetails
                              card:(AWXCard *)card
            shouldStoreCardDetails:(BOOL)storeCard
                             error:(NSString **)error {
    AWXPlaceDetails *validatedBilling;
    if (self.isBillingInformationRequired && placeDetails == nil) {
        if (error != NULL) {
            *error = NSLocalizedString(@"No billing address provided.", nil);
        }

        return NO;
    } else if (self.isBillingInformationRequired) {
        NSString *billingValidationError;
        validatedBilling = [self validatedBillingDetails:placeDetails error:&billingValidationError];
        if (validatedBilling == nil) {
            if (error != NULL && billingValidationError != nil) {
                *error = billingValidationError;
            }

            return NO;
        }
    }

    NSString *cardValidationError;
    AWXCard *validatedCard = [self validatedCardDetails:card error:&cardValidationError];
    if (validatedCard == nil) {
        if (error != NULL && cardValidationError != nil) {
            *error = cardValidationError;
        }

        return NO;
    }

    [provider confirmPaymentIntentWithCard:validatedCard billing:validatedBilling saveCard:storeCard];

    return YES;
}

- (void)updatePaymentIntentId:(NSString *)paymentIntentId {
    [self.session updateInitialPaymentIntentId:paymentIntentId];
}

#pragma mark Private helper

- (AWXBrandType)cardBrandFromCardScheme:(AWXCardScheme *)cardScheme {
    NSString *cardName = cardScheme.name;
    if ([cardName isEqualToString:@"amex"]) {
        return AWXBrandTypeAmex;
    } else if ([cardName isEqualToString:@"mastercard"]) {
        return AWXBrandTypeMastercard;
    } else if ([cardName isEqualToString:@"visa"]) {
        return AWXBrandTypeVisa;
    } else {
        return AWXBrandTypeUnknown;
    }
}

@end

File: Airwallex/Card/Internal/AWXCardViewModel.swift
//
//  AWXCardViewModel.swift
//  Card
//
//  Created by Tony He (CTR) on 2024/7/25.
//  Copyright © 2024 Airwallex. All rights reserved.
//

import Foundation

@objc public protocol AWXCardviewModelDelegate {
    @objc optional func startLoading()
    @objc optional func stopLoading()
    @objc func shouldDismiss(completeStatus status: AirwallexPaymentStatus, error: (any Error)?)
    @objc optional func didCompleteWithPaymentConsentId(_ id: String)
    @objc optional func shouldShowError(_ error: String)
    @objc func shouldPresent(_ controller: UIViewController?, forceToDismiss: Bool, withAnimation: Bool)
    @objc func shouldInsert(_ controller: UIViewController?)
}

@objcMembers
@objc
public class AWXCardViewModel: NSObject {
    public weak var delegate: AWXCardviewModelDelegate?
    public var isShowCardFlowDirectly = false

    public var ctaTitle: String {
        if session is AWXRecurringSession {
            return NSLocalizedString("Confirm", comment: "Confirm button title")
        } else {
            return NSLocalizedString("Pay", comment: "Pay button title")
        }
    }

    public let pageName: String = "card_payment_view"
    public var additionalInfo: [String: Any] {
        return ["supportedSchemes": supportedCardSchemes.map { $0.name }]
    }

    public private(set) var isReusingShippingAsBillingInformation: Bool = false
    public var isBillingInformationRequired: Bool {
        session?.isBillingInformationRequired ?? false
    }

    public var isCardSavingEnabled: Bool {
        session is AWXOneOffSession && session?.customerId() != nil
    }

    public var initialBilling: AWXPlaceDetails {
        session?.billing ?? AWXPlaceDetails(firstName: nil, lastName: nil, email: nil, dateOfBirth: nil, phoneNumber: nil, address: nil)
    }

    public var selectedCountry: AWXCountry?
    public var provider: AWXDefaultProvider?

    public let session: AWXSession?
    private var supportedCardSchemes: [AWXCardScheme] = []

    public init(session: AWXSession, supportedCardSchemes: [AWXCardScheme]) {
        self.session = session
        selectedCountry = AWXCountry.countryWithCode(session.billing?.address?.countryCode ?? "")
        isReusingShippingAsBillingInformation =
            session.billing != nil && session.isBillingInformationRequired
        self.supportedCardSchemes = supportedCardSchemes
    }

    public func setReusesShippingAsBillingInformation(_ reusesShippingAsBillingInformation: Bool)
        throws
    {
        if reusesShippingAsBillingInformation && session?.billing == nil {
            throw NSError.errorForAirwallexSDK(with: NSLocalizedString("No shipping address configured.", comment: ""))
        } else {
            isReusingShippingAsBillingInformation = reusesShippingAsBillingInformation
        }
    }

    public func makeBilling(
        firstName: String,
        lastName: String,
        email: String,
        phoneNumber: String,
        state: String,
        city: String,
        street: String,
        postcode: String
    ) -> AWXPlaceDetails {
        if isReusingShippingAsBillingInformation, let billing = session?.billing {
            return billing
        }

        let address = AWXAddress(countryCode: selectedCountry?.countryCode ?? "", city: city, street: street, state: state, postcode: postcode)
        return AWXPlaceDetails(firstName: firstName, lastName: lastName, email: email, dateOfBirth: nil, phoneNumber: phoneNumber, address: address)
    }

    public func makeCard(
        name: String,
        number: String,
        expiry: String,
        cvc: String
    ) -> AWXCard {
        let dates = expiry.split(separator: "/")
        return AWXCard(number: number.replacingOccurrences(of: " ", with: ""), expiryMonth: "\(dates.first ?? "")", expiryYear: "20\(dates.last ?? "")", name: name, cvc: cvc, bin: nil, last4: nil, brand: nil, country: nil, funding: nil, fingerprint: nil, cvcCheck: nil, avsCheck: nil, numberType: nil)
    }

    public func makeDisplayedCardBrands() -> [AWXCardBrand] {
        return AWXAllCardBrand().filter { cardBrand in
            supportedCardSchemes.map { $0.name }.contains(cardBrand.rawValue)
        }
    }

    private func validatedBillingDetails(_ billing: AWXPlaceDetails, error: inout String?)
        -> AWXPlaceDetails?
    {
        if let validationError = billing.validateAndReturnError() {
            error = validationError
            return nil
        } else {
            return billing
        }
    }

    private func validatedCardDetails(_ card: AWXCard, error: inout String?) -> AWXCard? {
        if let validationError = card.validateAndReturnError() {
            error = validationError
            return nil
        } else {
            return card
        }
    }

    public func validationMessageFromCardNumber(_ cardNumber: String) -> String? {
        if !cardNumber.isEmpty {
            if AWXCardValidator.shared.isValidCardLength(cardNumber) {
                let cardName = AWXCardValidator.shared.brandForCardNumber(cardNumber)?.name ?? ""
                for cardScheme in supportedCardSchemes {
                    if cardScheme.name == cardName.lowercased() {
                        return nil
                    }
                }
                return NSLocalizedString("Card not supported for payment", comment: "")
            }
            return NSLocalizedString("Card number is invalid", comment: "")
        }
        return NSLocalizedString("Card number is required", comment: "")
    }

    public func preparedProviderWithDelegate() -> AWXCardProvider {
        let provider = AWXCardProvider(delegate: self, session: session ?? AWXSession())
        self.provider = provider
        return provider
    }

    public func actionProviderForNextAction(
        _: AWXConfirmPaymentNextAction, delegate: AWXProviderDelegate
    ) -> AWXDefaultActionProvider {
        return AWX3DSActionProvider(delegate: delegate, session: session ?? AWXSession())
    }

    public func confirmPayment(
        provider: AWXCardProvider, billing placeDetails: AWXPlaceDetails?, card: AWXCard,
        shouldStoreCardDetails storeCard: Bool
    ) throws {
        var validatedBilling: AWXPlaceDetails?
        if isBillingInformationRequired, placeDetails == nil {
            throw NSError.errorForAirwallexSDK(with: NSLocalizedString("No billing address provided.", comment: ""))
        } else if isBillingInformationRequired, let placeDetails {
            var billingValidationError: String?
            validatedBilling = validatedBillingDetails(placeDetails, error: &billingValidationError)
            if validatedBilling == nil {
                throw NSError.errorForAirwallexSDK(with: billingValidationError ?? "")
            }
        }

        var cardValidationError: String?
        if let validatedCard = validatedCardDetails(card, error: &cardValidationError) {
            provider.confirmPaymentIntent(with: validatedCard, billing: validatedBilling, saveCard: storeCard)
        } else {
            throw NSError.errorForAirwallexSDK(with: cardValidationError ?? "")
        }
    }

    private func updatePaymentIntentId(_: String) {}

    private func cardBrandFromCardScheme(_ cardScheme: AWXCardScheme) -> AWXCardBrand {
        return AWXCardBrand(rawValue: cardScheme.name)
    }
}

extension AWXCardViewModel: AWXProviderDelegate {
    public func providerDidStartRequest(_: AWXDefaultProvider) {
        logMessage("providerDidStartRequest:")
        delegate?.startLoading?()
    }

    public func providerDidEndRequest(_: AWXDefaultProvider) {
        logMessage("providerDidEndRequest:")
        delegate?.stopLoading?()
    }

    public func provider(
        _: AWXDefaultProvider, didCompleteWith status: AirwallexPaymentStatus,
        error: (any Error)?
    ) {
        logMessage(
            "provider:didCompleteWithStatus:error: \(status)  \(error?.localizedDescription ?? "")")

        delegate?.shouldDismiss(completeStatus: status, error: error)
    }

    public func provider(_: AWXDefaultProvider, didCompleteWithPaymentConsentId id: String) {
        delegate?.didCompleteWithPaymentConsentId?(id)
    }

    public func provider(
        _: AWXDefaultProvider, didInitializePaymentIntentId paymentIntentId: String
    ) {
        logMessage("provider:didInitializePaymentIntentId:  \(paymentIntentId)")
        updatePaymentIntentId(paymentIntentId)
    }

    public func provider(
        _: AWXDefaultProvider, shouldHandle nextAction: AWXConfirmPaymentNextAction
    ) {
        logMessage(
            "provider:shouldHandleNextAction:  type:\(nextAction.type ?? ""), stage: \(nextAction.stage ?? "")")
        let actionProvider = actionProviderForNextAction(nextAction, delegate: self)
        actionProvider.handle(nextAction)
        provider = actionProvider
    }

    public func provider(
        _: AWXDefaultProvider, shouldPresent controller: UIViewController?,
        forceToDismiss: Bool, withAnimation: Bool
    ) {
        delegate?.shouldPresent(controller, forceToDismiss: forceToDismiss, withAnimation: withAnimation)
    }

    public func provider(_: AWXDefaultProvider, shouldInsert controller: UIViewController?) {
        delegate?.shouldInsert(controller)
    }
}
